devito/passes/iet/mpi.py

```diff
@@ -115,65 +100,139 @@ def rule1(dep, candidates, loc_dims):
     # Post-process analysis
     mapper = {i: HaloSpot(i._rebuild(), HaloScheme.union(hss))
               for i, hss in imapper.items()}
+
     mapper.update({i: i.body if hs.is_void else i._rebuild(halo_scheme=hs)
                    for i, hs in hsmapper.items()})
 
     # Transform the IET hoisting/dropping HaloSpots as according to the analysis
     iet = Transformer(mapper, nested=True).visit(iet)
 
     # Clean up: de-nest HaloSpots if necessary
-    mapper = {}
-    for hs in FindNodes(HaloSpot).visit(iet):
-        if hs.body.is_HaloSpot:
-            halo_scheme = HaloScheme.union([hs.halo_scheme, hs.body.halo_scheme])
-            mapper[hs] = hs._rebuild(halo_scheme=halo_scheme, body=hs.body.body)
-    iet = Transformer(mapper, nested=True).visit(iet)
+    iet = denest_halospots(iet)
 
     return iet
 
 
-def _merge_halospots(iet):
+def _hoist_and_drop(iet):
```

Comment:
it looks really complex, I hoped we could get away with something easier. It's a pity we have to add this much complexity and so much overlap with either hoist_halospot and merge_halospot